storage/ksql/src/main/java/io/apicurio/registry/storage/impl/ksql/sql/KafkaSqlSink.java

```diff
@@ -0,0 +1,264 @@
+                        value.getContentHash(), value.getCreatedBy(), value.getCreatedOn(),
+                        value.getMetaData(), globalIdGenerator);
+            case Update:
+                return sqlStore.updateArtifactWithMetadata(key.getArtifactId(), value.getArtifactType(),
+                        value.getContentHash(), value.getCreatedBy(), value.getCreatedOn(),
+                        value.getMetaData(), globalIdGenerator);
+            case Delete:
+                return sqlStore.deleteArtifact(key.getArtifactId());
+            case Clear:
+            default:
+                log.warn("Unsupported artifact message action: %s", key.getType().name());
+                throw new RegistryStorageException("Unsupported artifact message action: " + value.getAction());
+        }
+    }
+
+    /**
+     * Process a Kafka message of type "artifact rule".  This includes creating, updating, and deleting
+     * rules for a specific artifact.
+     * @param key
+     * @param value
+     */
+    private Object processArtifactRuleMessage(ArtifactRuleKey key, ArtifactRuleValue value) {
+        switch (value.getAction()) {
+            case Create:
+                // Note: createArtifactRuleAsync() must be called instead of createArtifactRule() because that's what 
+                // KafkaSqlRegistryStorage::createArtifactRuleAsync() expects (a return value)
+                return sqlStore.createArtifactRuleAsync(key.getArtifactId(), key.getRuleType(), value.getConfig());
+            case Update:
+                sqlStore.updateArtifactRule(key.getArtifactId(), key.getRuleType(), value.getConfig());
+                return null;
+            case Delete:
+                sqlStore.deleteArtifactRule(key.getArtifactId(), key.getRuleType());
+                return null;
+            case Clear:
+            default:
+                log.warn("Unsupported artifact rule message action: %s", key.getType().name());
+                throw new RegistryStorageException("Unsupported artifact-rule message action: " + value.getAction());
+        }
+    }
+
+    /**
+     * Process a Kafka message of type "artifact version".  This includes:
+     * 
+     *  - Updating version meta-data and state
+     *  - Deleting version meta-data and state
+     *  - Deleting an artifact version
+     * 
+     * @param key
+     * @param value
+     */
+    private Object processArtifactVersion(ArtifactVersionKey key, ArtifactVersionValue value) {
+        switch (value.getAction()) {
+            case Update:
+                sqlStore.updateArtifactVersionMetaDataAndState(key.getArtifactId(), key.getVersion(), value.getMetaData(), value.getState());
+                return null;
+            case Delete:
+                sqlStore.deleteArtifactVersion(key.getArtifactId(), key.getVersion());
+                return null;
+            case Clear:
+                sqlStore.deleteArtifactVersionMetaData(key.getArtifactId(), key.getVersion());
```

Comment:
I was not aware we had overlapping concepts , having `delete` and `clear` is not perfect as the meaning is similar but I guess it's ok if we have no other option

Reply:
What about joining both operations into a single one? Maybe I am missing something, but if the version is deleted, then the metadata can be safely deleted as well.

Reply:
Yeah I didn't love this solution, but I didn't have a better one.  I'm open to other options.  The semantics here are actually:

1) Delete == delete the **artifact version**
2) Clear == delete the artifact version **metadata**

They are different actions associated with the same message key, which is important IMO for log compaction.  If you can think of either a better action name, or another approach, I'm happy to hear it.